Stash integration (webhooks support and PR reporter)

EmilioPeJu opened the pull request:
# Main Features
- Support Bitbucket server webhook events 
- Send customized comments to a pull request in Bitbucket server

## Notes
- The last [pull request](https://github.com/dls-controls/buildbot/pull/2) was closed in order to avoid confusion
- I have squashed all the little commits

## Contributor Checklist:
* [x] I have updated the unit tests
* [x] I have created a file in the master/buildbot/newsfragment directory (and read the README.txt in that directory)
* [x] I have updated the appropriate documentation

noc0lour committed:
steps: add Assert step to master steps

noc0lour committed:
process: add comparator functions to Property class

rjarry committed:
virtualenv: upgrade setuptools to avoid errors

This problem was fixed in commit d2b9a0b6eb0e ("virtualenv: upgrade
setuptools to avoid errors") but was re-introduced recently.

Fixes: 395425eeb1ff ("added optional VENV_PY_VERSION arg for virtualenv make target")

Signed-off-by: Robin Jarry <robin.jarry@6wind.com>

protatremy committed:
creating populate secrets step

protatremy committed:
Push the secrets in file to the worker and delete it steps

trefzer committed:
fix for doStepIf not being renderable

- evaluation of doStepIf should be done after rendering

fixes Issue #3188

trefzer committed:
Add unit test for rendering doStepIf

trefzer committed:
fix sorting of imports in test_process_buildstep

Commit:
Create rmfile and download file to worker functions

rkashapov committed:
Fix BuildbotService name

protatremy committed:
flake8 fix

pyup-bot committed:
Update astroid from 1.5.2 to 1.5.3

pyup-bot committed:
Update astroid from 1.5.2 to 1.5.3

pyup-bot committed:
Update botocore from 1.5.56 to 1.5.60

pyup-bot committed:
Update isort from 4.2.5 to 4.2.13

pyup-bot committed:
Update isort from 4.2.5 to 4.2.13

pyup-bot committed:
Update moto from 1.0.0 to 1.0.1

pyup-bot committed:
Update pyasn1-modules from 0.0.8 to 0.0.9

protatremy committed:
rename tests and add worker version 3.1

protatremy committed:
adding rmfile to spelling list

protatremy committed:
pylint

protatremy committed:
flake8 over indent fix

tardyp committed:
isort+autopep8 run

tardyp committed:
relnotes

tardyp committed:
make sure shareservice hash generation is sorted

tardyp committed:
Merge pull request #3214 from trefzer/fix_doStepIf

fix for doStepIf not being renderable

EmilioPeJu committed:
added bitbucket server webhooks support

The following events will be supported:
    - push
    - pull request created
    - pull request updated
    - pull request fulfilled
    - pull request rejected

tardyp committed:
Merge pull request #3294 from buildbot/pyup-scheduled-update-06-05-2017

Scheduled weekly dependency update for week 23

tardyp committed:
Merge pull request #3202 from protatremy/populate

Push secret to worker

EmilioPeJu committed:
added support for sending comments to a PR in bitbucket server

It's implemented in the class StashPRCommentPush

EmilioPeJu committed:
StashStatusPush: allow to use got_revision

tardyp committed:
fix gitlab to use giturlparse

tardyp committed:
relnotes

tardyp committed:
isort

tardyp committed:
Create gitlabreporter.feature

tardyp committed:
increase coverage for gitlab reporter

tardyp committed:
source steps: make name and codebase renderable

tardyp committed:
add relnotes

tardyp committed:
fix escape sequence for py3

tardyp committed:
add a note in doc about renderability of name

rodrigc committed:
Update to Twisted 17.5.0

nazarewk committed:
Add methods to integrate secret population steps

(creation and deletion)

nazarewk committed:
PY3: fix request.method being bytes instead of str

additionally create tuple to begin with instead of list then changing it
to tuple

tardyp committed:
Merge pull request #3312 from nazarewk/master

PY3: fix request.method being bytes instead of str

EmilioPeJu committed:
reorder imports

EmilioPeJu committed:
make StringIO import python 2 and 3 compatible

EmilioPeJu committed:
Fix flake8 code style errors

EmilioPeJu committed:
Make plugin name explicit

EmilioPeJu committed:
Remove deprecated function call

tardyp committed:
fix tests form mail

tardyp committed:
Merge pull request #3310 from rodrigc/twisted-17.5

Update to Twisted 17.5.0

EmilioPeJu committed:
hardcoded Content-Type constant

tardyp committed:
relnotes for 0.9.8

tardyp committed:
Merge pull request #3308 from tardyp/gitlaburlparse

Gitlaburlparse

tardyp committed:
Merge pull request #3309 from tardyp/renderablecodebase

Renderablecodebase

rodrigc committed:
For twisted.mail.smtp.ESMTPSenderFactory, ensure username and password are bytes.

This fixes Python 3.

rodrigc committed:
Merge pull request #3317 from rodrigc/esmtp-bytes

twisted.mail.smtp.ESMTPSenderFactory fix for Python 3.

rodrigc committed:
Issues should be opened on GitHub, not trac.

tardyp committed:
update contributing.rst to use github issues instead of trac

rodrigc committed:
Merge pull request #3318 from rodrigc/issues-fix

Issues should be opened on GitHub, not trac.

ulrikpedersen reviewed (approved):
Nick did all the hard review work and I'm happy that this PR appear coherent; reasonable number of commits and sensible commit comments.
Thanks!

nickbattam commented:
These changes have been submitted to bitbucket: https://github.com/buildbot/buildbot/pull/3314

tardyp committed:
Merge pull request #3079 from noc0lour/add_assert_renderable

steps: add assert renderable step to master steps

EmilioPeJu committed:
add ref_deleted event and support tag events

tardyp committed:
Merge pull request #3315 from tardyp/releasenote

relnotes for 0.9.8

tardyp committed:
release scripts improvements

EmilioPeJu committed:
rename Stash to BitbucketServer

Frodox committed:
sync pathes in tour vs firstrun

tardyp committed:
fix spellcheck with version

Frodox committed:
minor syntax formating

tardyp committed:
Merge pull request #3320 from tardyp/release

release scripts improvements

tardyp committed:
gitlab hook: support proper scheme for mergerequest reporting

Frodox committed:
minor highlight formating to Pull Request Template

tardyp committed:
refactor the hook system for better extensivity

A lot of this code is imho clunky, but for this commit, I chose to keep it like it.
we just embed the current code in classes, and declare them in the plugin system.

tardyp committed:
newsfragment

tardyp committed:
Merge pull request #3326 from Frodox/patch-2

minor highlight formating to Pull Request Template

tardyp committed:
Merge pull request #3323 from Frodox/frodox-fix-tutorial

[docs] Sync pathes in tour vs firstrun

tardyp committed:
Merge pull request #3322 from Frodox/patch-1

minor syntax formating

tardyp committed:
fix relnote link

tardyp committed:
fix some py3 issues

tardyp committed:
allow getChange to return a deferred

tardyp committed:
spelling

tardyp committed:
Merge pull request #3282 from rkashapov/fix_service_name

Fix wrong GerritStatusPush.name attribute

tardyp committed:
Gitlab -> GitLab

EmilioPeJu committed:
Fix documentation references to renamed BitbucketServer classes

tardyp committed:
Merge pull request #3324 from tardyp/gitlab_mr

gitlab hook: support proper scheme for mergerequest reporting

tardyp committed:
flake8 fix

tardyp committed:
use pep440 for our intermediate version number

tardyp committed:
datetime version fallback

tardyp committed:
Merge pull request #3329 from tardyp/pep440

Pep440 intermediate versions

Commit:
adding test and doc

Commit:
pylint fix

Frodox committed:
[docs]: fix syntax and links in JanitorConfigurator

tardyp committed:
Merge pull request #3331 from Frodox/frodox-fix-janitor-docs

[docs]: fix syntax and links in JanitorConfigurator

noc0lour committed:
cmake: add docs for renderables in options/definitions

tardyp committed:
Merge pull request #3332 from noc0lour/add_cmake_docs

cmake: add docs for renderables in options/definitions

rjarry committed:
db: allow using in-memory sqlite uri

When using an in-memory sqlite database (db_url='sqlite://'),
everything is reset at each process restart. This causes the following
error:

  Setting up database with URL 'sqlite:'
  The Buildmaster database needs to be upgraded before this version of
  buildbot can run.  Use the following command-line

      buildbot upgrade-master ...

  to upgrade the database, and try starting the buildmaster again.  You may
  want to make a backup of your buildmaster before doing so.

Since there is no data persistence, the upgrade-master command would
work but everything would be lost when restarting the server.

Handle this special case and initialize the db model at database
connection to allow startup to continue.

Signed-off-by: Robin Jarry <robin.jarry@6wind.com>

Foxboron committed:
Make UserPasswordAuth string casting aware of lists

Signed-off-by: Morten Linderud <morten@linderud.pw>

tardyp committed:
Merge pull request #3335 from Foxboron/foxboron/fix-userpasswordauth-list

Make UserPasswordAuth string casting aware of lists

EmilioPeJu committed:
used MessageFormatter in order to render the comment text

pyup-bot committed:
Update autobahn from 17.5.1 to 17.6.1

pyup-bot committed:
Update botocore from 1.5.60 to 1.5.70

pyup-bot committed:
Update isort from 4.2.13 to 4.2.15

pyup-bot committed:
Update isort from 4.2.13 to 4.2.15

pyup-bot committed:
Update lz4 from 0.9.4 to 0.10.0

pyup-bot committed:
Update pathlib2 from 2.2.1 to 2.3.0

pyup-bot committed:
Update requests from 2.17.3 to 2.18.1

pyup-bot committed:
Update txaio from 2.7.1 to 2.8.0

pyup-bot committed:
Update zope.interface from 4.4.1 to 4.4.2

pyup-bot committed:
Update zope.interface from 4.4.1 to 4.4.2

pyup-bot committed:
Update buildbot-www from 0.9.7 to 0.9.8

tardyp committed:
Merge pull request #3333 from rjarry/sqlite-memory

db: allow using in-memory sqlite uri

rjarry committed:
www: add grid_view plugin

Add a grid view plugin.

It is possible to display changes in all branches or restrict to a
specific branch. It is also possible to filter builders by clicking on
builder tags.

In the settings page, the user can specify how many changes are
displayed at max plus other more internal stuff.

Signed-off-by: Robin Jarry <robin.jarry@6wind.com>

chrwen-omicron committed:
Use string_types to validate the command in the ShellArg class.

Commit:
test config methods are factorized

Commit:
secrets providers file and vault are added to the setup list

nazarewk committed:
#3339 NOP if RemoteCommand.build is None

rjarry committed:
www/base: add compact option to changedetails directive

Allow hiding the avatar and (time ago) suffix in directive details.
Example:

    changedetails(change="ch", compact="true")

Signed-off-by: Robin Jarry <robin.jarry@6wind.com>

tardyp committed:
cleanup reclaim and unclaimexpire

This is not needed anymore in nine, as we now have master expiration, which does the same thing

krf committed:
Minor: Fix casing of property source

Consistency++

rjarry committed:
gitignore: ignore .venv folder

Signed-off-by: Robin Jarry <robin.jarry@6wind.com>

rodrigc committed:
Python 3.4 doesn't support PEP 461 bytes formatting.

tardyp committed:
removed unused import

EmilioPeJu committed:
implemented assertNotLogged in LoggingMixin

Commit:
declare providers in packages

rkashapov committed:
Remove statement with no effect

EmilioPeJu committed:
remove unnecessary header check

EmilioPeJu committed:
extend test coverage

EmilioPeJu committed:
fix problems pointed by the build reports

EmilioPeJu committed:
more build report issues, mainly reorder of imports

EmilioPeJu committed:
Remove PR comment from log message

This is not required and causes compatibility issues with unicode handling in twisted 14.0.2

opalmer committed:
_wait_for_request() should be using request_status

_wait_for_request() is using request.status when it should be
using request_status instead. Otherwise an error will be raised
when the log message attempts to format request.status.

EmilioPeJu committed:
update log message in the tests not expecting PR comment

opalmer committed:
updating news file

tardyp committed:
Merge pull request #3358 from opalmer/master

_wait_for_request() should be using request_status

tardyp committed:
Merge pull request #3354 from rkashapov/remove_no_effect_statement

Remove statement with no effect

tardyp committed:
Merge pull request #3350 from rodrigc/bytes-py34

Python 3.4 doesn't support PEP 461 bytes formatting.

tardyp committed:
Merge pull request #3345 from rjarry/grid-view

www: add grid_view plugin

tardyp committed:
Merge pull request #3346 from tardyp/cleanup_builder_start

cleanup reclaim and unclaimexpire

tardyp committed:
Merge pull request #3344 from protatremy/setup

secrets providers file and vault are added to the setup list

tardyp committed:
Merge pull request #3347 from krf/fix-property-source

Minor: Fix casing of property source

tardyp committed:
Merge pull request #3337 from buildbot/pyup-scheduled-update-06-19-2017

Scheduled weekly dependency update for week 25

tardyp committed:
Merge pull request #3327 from tardyp/hook_services

Hook plugins

tardyp committed:
Merge pull request #3330 from protatremy/withsecret

Add methods to integrate secret population steps

tardyp committed:
Merge pull request #3338 from chrwen-omicron/shellarg_command_string

Use string_types to validate the command in the ShellArg class.

tardyp committed:
Merge pull request #3343 from nazarewk/master

#3339 NOP if RemoteCommand.build is None

EmilioPeJu committed:
integrate hook to the plugin system

according to pr #3327

EmilioPeJu committed:
use a more standard way to report

EmilioPeJu committed:
set revision to source branch commit on the pull request events

EmilioPeJu committed:
mainly build report issues related to the style
